### PR TITLE
Add recipe for mosh - mobile shell

### PR DESCRIPTION
**Pull request by gokceneraslan**
Mosh (mobile shell)

> Remote terminal application that allows roaming, supports intermittent connectivity, and provides intelligent local echo and line editing of user keystrokes.
> 
> Mosh is a replacement for SSH. It's more robust and responsive, especially over Wi-Fi, cellular, and long-distance links.
> 

Packaging mosh for conda is highly important because `mosh` is required to be installed on both server and client side. However users may not have the root account on servers and manual compilation of mosh can be tricky especially if header packages or protobuf compiler are not installed on server. 

Aim of this packages is to make things easier e.g. users can run just `conda install -c conda-forge mosh` and they're done.

### CODE REVIEW

File: recipes/mosh/build.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    export CC=clang
+    export CXX=clang++
+    export CXXFLAGS="-mmacosx-version-min=${MACOSX_VERSION_MIN}"
+    export CXXFLAGS="${CXXFLAGS} -stdlib=libc++ -std=c++11 -I${PREFIX}/include"
+    export LIBS="-lc++"
+elif [ "$(uname)" == "Linux" ];
+then
+    export CC=gcc
+    export CXX=g++
+fi
```
Comment (scopatz) [inline recipes/mosh/build.sh]: Most of the above might not be needed if toolchain is used.

File: recipes/mosh/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  url: https://mosh.org/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - gcc  # [linux]
```
Comment (gokceneraslan) [inline recipes/mosh/meta.yaml]: Updated, thanks.

Comment (scopatz) [inline recipes/mosh/meta.yaml]: np

**Comment by gokceneraslan** [comment]
Since I switched to `toolchain`, it doesn't compile on Linux any more. I cannot reproduce the `zlib not found` error. Also conda-build is downgraded to 1.21.14 for some reason. Any ideas?

File: recipes/mosh/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    - protobuf
+    - ncurses
+    - zlib
+    - openssl
+  run:
+    - protobuf
+    - ncurses
+    - zlib
+    - openssl
+    - perl
```
Comment (scopatz) [inline recipes/mosh/meta.yaml]: Some of these need to be pinned to the correct version: https://github.com/conda-forge/conda-forge.github.io/blob/master/scripts/pin_the_slow_way.py#L28 for example, `zlib 1.2.*`

File: recipes/mosh/build.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# See https://github.com/conda-forge/toolchain-feedstock/pull/11
+export CXXFLAGS="${CXXFLAGS} -I${PREFIX}/include `pkg-config --cflags zlib`"
+export LDFLAGS="${LDFLAGS} -L${PREFIX}/lib `pkg-config --libs zlib`"
```
Comment (scopatz) [inline recipes/mosh/build.sh]: Would you mind using `$(...)` instead of  backticks, please

Comment (gokceneraslan) [inline recipes/mosh/build.sh]: ok.

File: recipes/mosh/meta.yaml (hunk 3 above)
Comment (gokceneraslan) [inline recipes/mosh/meta.yaml]: Could you please elaborate? I didn't quite get the purpose of this list. What is this script and why are there a hard-coded list of packages?

Comment (scopatz) [inline recipes/mosh/meta.yaml]: It is a list of ABI compatability issues. Adhering to these versions means all packages in conda-forge can interop.

**Comment by gokceneraslan** [comment]
Any comments @msarahan @isuruf ?

File: recipes/mosh/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    - pkg-config
+    - protobuf
+    - ncurses 5.9*
+    - zlib 1.2.*
+    - openssl 1.0.*
+  run:
+    - protobuf
+    - ncurses 5.9*
+    - zlib 1.2.*
+    - openssl 1.0.*
```
Comment (isuruf) [inline recipes/mosh/meta.yaml]: To make sure `openssl` is picked add `--with-crypto-library=openssl` to `./configure` call

File: recipes/mosh/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  build:
+    - toolchain
+    - pkg-config
+    - protobuf
+    - ncurses 5.9*
+    - zlib 1.2.*
+    - openssl 1.0.*
+  run:
+    - protobuf
+    - ncurses 5.9*
```
Comment (isuruf) [inline recipes/mosh/meta.yaml]: To make sure ncurses is used add, `--with-ncurses --with-curses=$PREFIX` to `./configure` call

File: recipes/mosh/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    - perl
+
+test:
+  commands:
+    - mosh-server -h
+
+about:
+  home: https://mosh.org/
+  license: GPLv3+
+  license_family: GPL3
```
Comment (isuruf) [inline recipes/mosh/meta.yaml]: You can add the license file by adding, `license_file: COPYING`

File: recipes/mosh/build.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# See https://github.com/conda-forge/toolchain-feedstock/pull/11
+export CXXFLAGS="${CXXFLAGS} -I${PREFIX}/include $(pkg-config --cflags zlib)"
+export LDFLAGS="${LDFLAGS} -L${PREFIX}/lib $(pkg-config --libs zlib)"
```
Comment (isuruf) [inline recipes/mosh/build.sh]: You should use either `-L${PREFIX}/lib` or `$(pkg-config --libs-only-L zlib)`. Both are equivalent.
`$(pkg-config --libs zlib)` on the other hand adds `-lz` as well, which is unnecessary to find the libraries.

File: recipes/mosh/meta.yaml (hunk 6 above)
Comment (gokceneraslan) [inline recipes/mosh/meta.yaml]: Done.

File: recipes/mosh/meta.yaml (hunk 5 above)
Comment (gokceneraslan) [inline recipes/mosh/meta.yaml]: Done.

File: recipes/mosh/meta.yaml (hunk 7 above)
Comment (gokceneraslan) [inline recipes/mosh/meta.yaml]: OK.

**Comment by isuruf** [comment]
Looks good to me

File: recipes/mosh/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    - protobuf
+    - ncurses 5.9*
+    - zlib 1.2.*
+    - openssl 1.0.*
+  run:
+    - protobuf
+    - ncurses 5.9*
+    - zlib 1.2.*
+    - openssl 1.0.*
+    - perl
```
Comment (jakirkham) [inline recipes/mosh/meta.yaml]: What's `perl` needed for? If it is required to run the shell, then it should be in `requirements/build` also.

**Comment by gokceneraslan** [comment]
It's because `mosh` command (https://github.com/mobile-shell/mosh/blob/master/scripts/mosh.pl) is a simple Perl script.

 It's not required to build the shell, why would we write it as a build dependency?

**Comment by scopatz** [comment]
LGTM too, Thanks

File: recipes/mosh/meta.yaml (hunk 9 above)
Comment (scopatz) [inline recipes/mosh/meta.yaml]: yeah, I think mosh is weird in that the server is in C++ but the front end app is in perl

Comment (scopatz) [inline recipes/mosh/meta.yaml]: So perl is not a build dep

Comment (scopatz) [inline recipes/mosh/meta.yaml]: Ahh yeah, this is probably worth fixing in the feedstock.

Comment (scopatz) [inline recipes/mosh/meta.yaml]: Also, it would be nice to have a mention in the docs about this.  Maybe "packages that have to be build deps"?

**Comment by gokceneraslan** [comment]
Thank you.

**Pull request merged**